File: +ndi/+daq/reader.m
classdef reader < ndi.ido & ndi.documentservice
% NDI_DAQREADER - A class for objects that read samples for NDI_DAQSYSTEM objects
%
%
	properties (GetAccess=public, SetAccess=protected)
	end

	methods

		function obj = reader(varargin)
			% ndi.daq.reader - create a new ndi.daq.reader object
			%
			%  OBJ = ndi.daq.reader()
			%  
			%  Creates an ndi.daq.reader. 
			%
			%  OBJ = ndi.daq.reader(NDI_SESSION_OBJ, NDI_DOCUMENT_OBJ)
			%    
			%  Creates an ndi.daq.reader from an NDI_DOCUMENT_OBJ.
			%
			%  ndi.daq.reader has mostly abstract methods, it is made to be overriden.
			%
			obj = obj@ndi.ido();

			loadfromfile = 0;

			if nargin==2 & isa(varargin{1},'ndi.session') & isa(varargin{2},'ndi.document'),
				obj.identifier = varargin{2}.document_properties.base.id;
			elseif nargin>=2,
				if ischar(varargin{2}), % it is a command
					loadfromfile = 1;
					filename = varargin{1};
					if ~strcmp(lower(varargin{2}), lower('OpenFile')),
						error(['Unknown command.']);
					end
				end;
			end;

			if loadfromfile,
				error(['Load from file no longer supported.']);
			end
		end % ndi.daq.reader

		function d = getingesteddocument(ndi_daqreader_mfdaq_obj, epochfiles)
			% GETINGESTEDDOCUMENT - return the ndi.document with the data for an ingested epoch
			%
			% D = GETINGESTEDDOCUMENT(NDI_DAQREADER_MFDAQ_OBJ, EPOCHFILES)
			%
			% Returns the document D that contains the ingested ndi.daq.reader.mfdaq data for
			% the epoch described by EPOCHFILES. EPOCHFILES should be an ingested epoch.
			%
				assert(ndi.file.navigator.isingested(epochfiles),...
					'This function is only applicable to ingested EPOCHFILES.');

				error('needs development');			

		end % getingesteddocument();


		% EPOCHSET functions, although this object is NOT an EPOCHSET object

		function ec = epochclock(ndi_daqreader_obj, epochfiles)
			% EPOCHCLOCK - return the ndi.time.clocktype objects for an epoch
			%
			% EC = EPOCHCLOCK(NDI_DAQREADER_OBJ, EPOCHFILES)
			%
			% Return the clock types available for this epoch as a cell array
			% of ndi.time.clocktype objects (or sub-class members).
			%
			% For the generic ndi.daq.reader, this returns a single clock
			% type 'no_time';
			%
			% See also: ndi.time.clocktype
			%
				ec = {ndi.time.clocktype('no_time')};
		end % epochclock

		function ec = epochclock_ingested(ndi_daqreader_obj, epochfiles)
			% EPOCHCLOCK_INGESTED - return the ndi.time.clocktype objects for an ingested epoch
			%
			% EC = EPOCHCLOCK_INGESTED(NDI_DAQREADER_OBJ, EPOCHFILES)
			%
			% Return the clock types available for this epoch as a cell array
			% of ndi.time.clocktype objects (or sub-class members).
			%
			% See also: ndi.time.clocktype
			%
				error('needs development.');
		end % epochclock_ingested

		function t0t1 = t0_t1(ndi_epochset_obj, epochfiles)
			% T0T1 - return the t0_t1 (beginning and end) epoch times for an epoch
			%
			% T0T1 = T0_T1(NDI_EPOCHSET_OBJ, EPOCH_NUMBER)
			%
			% Return the beginning (t0) and end (t1) times of the epoch EPOCH_NUMBER
			% in the same units as the ndi.time.clocktype objects returned by EPOCHCLOCK.
			%
			% The abstract class always returns {[NaN NaN]}.
			%
			% See also: ndi.time.clocktype, EPOCHCLOCK
			%
				t0t1 = {[NaN NaN]};
		end % t0t1

		function t0t1 = t0_t1_ingested(ndi_epochset_obj, epochfiles)
			% T0T1_INGESTED - return the t0_t1 (beginning and end) epoch times for an ingested epoch
			%
			% T0T1 = T0_T1_INGESTED(NDI_EPOCHSET_OBJ, EPOCH_NUMBER)
			%
			% Return the beginning (t0) and end (t1) times of the epoch EPOCH_NUMBER
			% in the same units as the ndi.time.clocktype objects returned by EPOCHCLOCK.
			%
			% The abstract class always returns {[NaN NaN]}.
			%
			% See also: ndi.time.clocktype, EPOCHCLOCK
			%
				error('needs development.');
		end % t0t1

		function [b,msg] = verifyepochprobemap(ndi_daqreader_obj, epochprobemap, epochfiles)
			% VERIFYEPOCHPROBEMAP - Verifies EPOCHPROBEMAP is compatible with device and data on disk
			%
			% B = VERIFYEPOCHPROBEMAP(NDI_DAQREADER_OBJ, EPOCHPROBEMAP, NUMBER)
			%
			% Examines the ndi.epoch.epochprobemap_daqsystem EPOCHPROBEMAP and determines if it is
			% valid for the given device epoch NUMBER.
			%
			% For the abstract class ndi.daq.reader, EPOCHPROBEMAP is always valid as long as
			% EPOCHPROBEMAP is an ndi.epoch.epochprobemap_daqsystem object.
			%
			% See also: ndi.daq.reader, ndi.epoch.epochprobemap_daqsystem
				msg = '';
				b = isa(epochprobemap, 'ndi.epoch.epochprobemap_daqsystem');
				if ~b,
					msg = 'epochprobemap is not a member of the class ndi.epoch.epochprobemap_daqsystem; it must be.';
					return;
				end;

				for i=1:numel(epochprobemap),
					try,
						thedevicestring = ndi_daqreaderstring(epochprobemap(i).devicestring);
					catch,
						b = 0;
						msg = ['Error evaluating devicestring ' epochprobemap(i).devicestring '.'];
                                        end
                                end
		end % verifyepochprobemap

		function d = ingest_epochfiles(ndi_daqreader_obj, epochfiles)
			% INGEST_EPOCHFILES - create a document that describes data read by an ndi.daq.reader
			%
			% D = INGEST_EPOCHFILES(NDI_DAQREADER_OBJ, EPOCHFILES)
			%
			% Creates an ndi.document of type 'daqreader_epochdata_ingested' that contains the data
			% for an ndi.daq.reader object. The document D is not added to any database.
			%
			% Example:
			%    D = mydaqreader.ingest_epochfiles(epochfiles);
			
				d = ndi.document('daqreader_epochdata_ingested');
				d = d.set_dependency_value('daqreader_id',ndi_daqreader_obj.id());

		end; % ingest_epochfiles()

		function b = eq(ndi_daqreader_obj1, ndi_daqreader_obj2)
			% EQ - tests whether 2 ndi.daq.reader objects are equal
			%
			% B = EQ(NDI_DAQREADER_OBJ1, NDI_DAQREADER_OBJ2)
			%
			% Examines whether or not the ndi.daq.reader objects are equal.
			%
				b = strcmp(class(ndi_daqreader_obj1),class(ndi_daqreader_obj2));
				b = b & strcmp(ndi_daqreader_obj1.id(), ndi_daqreader_obj2.id());
		end; % eq()
		
		%% functions that override ndi.documentservice

		function ndi_document_obj = newdocument(ndi_daqreader_obj)
			% NEWDOCUMENT - create a new ndi.document for an ndi.daq.reader object
			%
			% DOC = NEWDOCUMENT(NDI_DAQREADER_OBJ)
			%
			% Creates an ndi.document object DOC that represents the
			%    ndi.daq.reader object. 
				ndi_document_obj = ndi.document('daq/daqreader.json',...
					'daqreader.ndi_daqreader_class',class(ndi_daqreader_obj),...
					'base.id', ndi_daqreader_obj.id(),...
					'base.session_id',ndi.session.empty_id());
		end; % newdocument()

		function sq = searchquery(ndi_daqreader_obj)
			% SEARCHQUERY - create a search for this ndi.daq.reader object
			%
			% SQ = SEARCHQUERY(NDI_DAQREADER_OBJ)
			%
			% Creates a search query for the ndi.daq.reader object. 
			% 
				sq = ndi.query('base.id','exact_string', ndi_daqreader_obj.id(), '');
		end; % searchquery()

	end % methods
		
end % ndi.daq.reader classdef

File: ndi-python/ndi/daq/reader.py
"""
NDI DAQ Reader - Base class for reading data from acquisition systems.
"""

from typing import List, Dict, Any, Optional, Tuple
from abc import ABC, abstractmethod

from ..ido import IDO
from ..document import Document
from ..query import Query


class Reader(IDO, ABC):
    """
    NDI DAQ Reader - abstract base class for data acquisition readers.

    The Reader class provides the interface for reading data from various
    data acquisition systems. Specific readers (e.g., Intan, Blackrock)
    inherit from this class and implement format-specific methods.

    Attributes:
        None in base class - subclasses may add format-specific attributes

    Examples:
        >>> # Subclass example (actual implementations in reader/ subdirectory)
        >>> from ndi.daq.reader import MFDAQReader
        >>> # reader = MFDAQReader()
    """

    def __init__(self):
        """Create a new DAQ Reader."""
        super().__init__()

    def epochclock(self, epochfiles: List[str]) -> List[Any]:
        """
        Return the clock types available for this set of epoch files.

        Args:
            epochfiles: List of epoch file paths

        Returns:
            List of ClockType objects

        Notes:
            Base class returns 'no_time' clock type.
            Subclasses should override to provide actual clock types.
        """
        from ..time import ClockType
        return [ClockType('no_time')]

    def epochclock_ingested(self, epochfiles: List[str], session) -> List[Any]:
        """
        Return clock types for ingested epoch files.

        Args:
            epochfiles: List of epoch file paths
            session: NDI Session object

        Returns:
            List of ClockType objects from ingested document
        """
        d = self.getingesteddocument(epochfiles, session)
        if d is None:
            return self.epochclock(epochfiles)

        ec_list = d.document_properties.get('daqreader_epochdata_ingested', {})\
                                       .get('epochtable', {})\
                                       .get('epochclock', [])

        if not ec_list:
            return self.epochclock(epochfiles)

        from ..time import ClockType
        result = []
        for ec in ec_list:
            if isinstance(ec, str):
                result.append(ClockType(ec))
            else:
                result.append(ec)
        return result

    def t0_t1(self, epochfiles: List[str]) -> List[List[float]]:
        """
        Return beginning and end times for the epoch.

        Args:
            epochfiles: List of epoch file paths

        Returns:
            List of [t0, t1] pairs for each clock type

        Notes:
            Base class returns [[NaN, NaN]].
            Subclasses should override to provide actual times.
        """
        import math
        return [[math.nan, math.nan]]

    def t0_t1_ingested(self, epochfiles: List[str], session) -> List[List[float]]:
        """
        Return beginning and end times for ingested epoch files.

        Args:
            epochfiles: List of epoch file paths
            session: NDI Session object

        Returns:
            List of [t0, t1] pairs from ingested document
        """
        d = self.getingesteddocument(epochfiles, session)
        if d is None:
            return self.t0_t1(epochfiles)

        t0t1 = d.document_properties.get('daqreader_epochdata_ingested', {})\
                                    .get('epochtable', {})\
                                    .get('t0_t1', [[]])

        # Handle conversion from JSON (may not be in cell array format)
        if not isinstance(t0t1, list):
            return [[t0t1[0], t0t1[1]]]
        elif len(t0t1) > 0 and not isinstance(t0t1[0], list):
            return [t0t1]

        return t0t1 if t0t1 else [[]]

    def verifyepochprobemap(self, epochprobemap: Any, epochfiles: List[str]) -> Tuple[bool, str]:
        """
        Verify that an epoch probe map is compatible with the data.

        Args:
            epochprobemap: Epoch probe map to verify
            epochfiles: List of epoch file paths

        Returns:
            Tuple of (is_valid, error_message)

        Notes:
            Base class only checks if epochprobemap is of correct type.
            Subclasses should implement more specific validation.
        """
        # TODO: Check isinstance of epochprobemap_daqsystem when implemented
        # For now, basic validation
        msg = ''
        b = isinstance(epochprobemap, (dict, list, type(None)))
        if not b:
            msg = 'epochprobemap must be dict, list, or None'
        return b, msg

    def ingest_epochfiles(self, epochfiles: List[str], epoch_id: str = '') -> Document:
        """
        Create a document that describes data read by this reader.

        Args:
            epochfiles: List of epoch file paths
            epoch_id: Epoch ID string

        Returns:
            Document of type 'daqreader_epochdata_ingested'

        Notes:
            The document is not added to any database.
            Subclasses should override to add format-specific data.
        """
        from ..time import ClockType

        # Get epoch clock and t0_t1
        ec = self.epochclock(epochfiles)
        ec_strings = [ct.type if isinstance(ct, ClockType) else str(ct) for ct in ec]

        t0t1 = self.t0_t1(epochfiles)

        # Convert t0_t1 to array format
        import numpy as np
        if len(t0t1) > 0 and isinstance(t0t1[0], list):
            t0t1_array = np.array(t0t1)
        else:
            t0t1_array = np.array([t0t1])

        daqreader_epochdata_ingested = {
            'epochtable': {
                'epochclock': ec_strings,
                't0_t1': t0t1_array.tolist()
            }
        }

        d = Document('daqreader_epochdata_ingested',
                    daqreader_epochdata_ingested=daqreader_epochdata_ingested)
        d = d.set_dependency_value('daqreader_id', self.id())

        return d

    def getingesteddocument(self, epochfiles: List[str], session) -> Optional[Document]:
        """
        Retrieve the ingested document for this epoch.

        Args:
            epochfiles: List of epoch file paths
            session: NDI Session object

        Returns:
            Ingested document, or None if not found

        Notes:
            Searches the session database for ingested epoch data.
        """
        if session is None:
            return None

        # Search for ingested document with this reader's ID
        # TODO: Also filter by epoch files when file tracking is implemented
        q = Query('daqreader_epochdata_ingested.daqreader_id', 'exact_string', self.id())

        try:
            docs = session.database_search(q)
            if len(docs) > 0:
                return docs[0]
        except:
            pass

        return None

    def ingested2epochs_t0t1_epochclock(self, session) -> Dict[str, Dict[str, Any]]:
        """
        Map ingested epochs to their t0_t1 and epochclock values.

        Args:
            session: NDI Session object

        Returns:
            Dictionary with 'epochclock' and 't0t1' subdictionaries,
            each mapping epoch_id to values

        Notes:
            Used by DAQ system to populate epoch tables from ingested data.
        """
        result = {
            'epochclock': {},
            't0t1': {}
        }

        if session is None:
            return result

        # Search for all ingested documents for this reader
        q = Query('daqreader_epochdata_ingested.daqreader_id', 'exact_string', self.id())

        try:
            docs = session.database_search(q)
            for doc in docs:
                # Extract epoch_id if available
                epoch_id = doc.document_properties.get('epochid', {}).get('epochid', '')
                if not epoch_id:
                    continue

                # Extract epochclock
                ec = doc.document_properties.get('daqreader_epochdata_ingested', {})\
                                           .get('epochtable', {})\
                                           .get('epochclock', [])
                if ec:
                    from ..time import ClockType
                    result['epochclock'][epoch_id] = [
                        ClockType(e) if isinstance(e, str) else e for e in ec
                    ]

                # Extract t0_t1
                t0t1 = doc.document_properties.get('daqreader_epochdata_ingested', {})\
                                              .get('epochtable', {})\
                                              .get('t0_t1', [])
                if t0t1:
                    result['t0t1'][epoch_id] = t0t1

        except Exception as e:
            # If search fails, return empty result
            pass

        return result

    def __eq__(self, other: 'Reader') -> bool:
        """
        Test whether two DAQ readers are equal.

        Args:
            other: Another Reader object

        Returns:
            True if same class and same ID
        """
        if not isinstance(other, Reader):
            return False
        return (type(self) == type(other) and self.id() == other.id())

    def newdocument(self) -> Document:
        """
        Create a new document for this DAQ reader.

        Returns:
            Document object representing this reader
        """
        doc = Document('daqreader',
                      daqreader_ndi_daqreader_class=f"{self.__class__.__module__}.{self.__class__.__name__}")
        doc.document_properties['base']['id'] = self.id()
        doc.document_properties['base']['session_id'] = ''  # No session for bare reader

        return doc

    def searchquery(self) -> Query:
        """
        Create a search query for this DAQ reader.

        Returns:
            Query object that searches by ID
        """
        return Query('base.id', 'exact_string', self.id())

    def __repr__(self) -> str:
        """String representation."""
        return f"{self.__class__.__name__}(id='{self.id()}')"

    def __str__(self) -> str:
        """String representation."""
        return self.__repr__()
